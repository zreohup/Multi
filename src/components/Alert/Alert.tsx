import { View, Text, Theme } from 'tamagui'
import React, { type ReactElement } from 'react'
import { SafeFontIcon } from '@/src/components/SafeFontIcon/SafeFontIcon'
import { IconName } from '@/src/types/iconTypes'
import { TouchableOpacity } from 'react-native'

export type AlertType = 'error' | 'warning' | 'info' | 'success'
export type AlertOrientation = 'left' | 'center' | 'right'

interface AlertProps {
  type: AlertType
  message: string | React.ReactNode
  info?: string | React.ReactNode
  iconName?: IconName
  displayIcon?: boolean
  fullWidth?: boolean
  endIcon?: React.ReactNode
  startIcon?: React.ReactNode
  onPress?: () => void
  testID?: string
  orientation?: AlertOrientation
}

const icons = {
  error: <SafeFontIcon testID="error-icon" name={'alert'} />,
  warning: (
    <View
      display="flex"
      alignItems="center"
      justifyContent="center"
      padding="$1"
      borderRadius={'$10'}
      backgroundColor="#FF8C00"
    >
      <SafeFontIcon testID="warning-icon" name={'alert'} color={'$colorContrast'} size={16} />
    </View>
  ),
  info: <SafeFontIcon testID="info-icon" name={'info'} />,
  success: <SafeFontIcon testID="success-icon" name={'check'} />,
}

const getAlertIcon = (type: AlertType, iconName?: IconName, displayIcon?: boolean): ReactElement | null => {
  if (!displayIcon) {
    return null
  }

  return iconName ? <SafeFontIcon testID={`${iconName}-icon`} name={iconName} /> : icons[type]
}

const getContainerAlignment = (orientation: AlertOrientation) => {
  switch (orientation) {
    case 'left':
      return 'flex-start'
    case 'right':
      return 'flex-end'
    case 'center':
    default:
      return 'center'
  }
}

const getContentAlignment = (orientation: AlertOrientation) => {
  switch (orientation) {
    case 'left':
      return 'flex-start'
    case 'right':
      return 'flex-start' // Still flex-start for content alignment, but container will be aligned right
    case 'center':
    default:
      return 'center'
  }
}

export const Alert = ({
  type,
  fullWidth = true,
  message,
  iconName,
  startIcon,
  endIcon,
  displayIcon = true,
  onPress,
  testID,
  info,
  orientation = 'center',
}: AlertProps) => {
  const Icon = getAlertIcon(type, iconName, displayIcon)
  const containerAlignment = getContainerAlignment(orientation)
  const contentAlignment = getContentAlignment(orientation)

  return (
    <Theme name={type}>
      <TouchableOpacity disabled={!onPress} onPress={onPress} testID={testID}>
        <View flexDirection="row" width="100%" justifyContent={containerAlignment}>
          <View
            alignItems="center"
            gap={'$3'}
            width={fullWidth ? '100%' : 'auto'}
            flexDirection="row"
            justifyContent={contentAlignment}
            backgroundColor="$background"
            padding="$2"
            borderRadius={'$2'}
          >
            {startIcon ? <View testID="alert-start-icon">{startIcon}</View> : Icon}

            <View gap={'$1'} flex={orientation !== 'center' ? 1 : undefined}>
              {typeof message === 'string' ? <AlertTitleStyled message={message} /> : message}
              {info && typeof info === 'string' ? (
                <Text fontSize={'$3'} fontFamily={'$body'}>
                  {info}
                </Text>
              ) : (
                info
              )}
            </View>

            {endIcon && <View testID="alert-end-icon">{endIcon}</View>}
          </View>
        </View>
      </TouchableOpacity>
    </Theme>
  )
}

export const AlertTitleStyled = ({ message }: { message: string }) => {
  if (typeof message === 'string') {
    return (
      <Text fontSize={'$4'} fontWeight={'600'} fontFamily={'$body'}>
        {message}
      </Text>
    )
  }

  return message
}
