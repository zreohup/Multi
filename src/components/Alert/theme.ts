import { tokens } from '@/src/theme/tokens'

export const alertTheme = {
  light_success: {
    background: tokens.color.successBackgroundLight,
    color: tokens.color.successMainLight,
    badgeBackground: tokens.color.successDarkLight,
    badgeTextColor: tokens.color.backgroundMainDark,
  },
  light_info: {
    background: tokens.color.infoBackgroundLight,
    color: tokens.color.infoMainLight,
  },
  light_warning: {
    background: tokens.color.warning1MainLight,
    color: tokens.color.warning1TextLight,
  },
  light_error: {
    background: tokens.color.error1MainLight,
    color: tokens.color.error1ContrastTextLight,
  },
  dark_success: {
    background: tokens.color.successBackgroundDark,
    color: tokens.color.successMainDark,
    badgeBackground: tokens.color.successDarkDark,
  },
  dark_info: {
    background: tokens.color.infoBackgroundDark,
    color: tokens.color.infoMainDark,
  },
  dark_warning: {
    background: tokens.color.warning1MainDark,
    color: tokens.color.warning1TextDark,
  },
  dark_error: {
    background: tokens.color.error1MainDark,
    color: tokens.color.error1ContrastTextDark,
  },
}
