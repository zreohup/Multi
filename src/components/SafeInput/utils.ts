export const getInputThemeName = (hasError?: boolean, hasSuccess?: boolean) => {
  if (hasError) {
    return 'error'
  }

  if (hasSuccess) {
    return 'success'
  }

  return 'default'
}
