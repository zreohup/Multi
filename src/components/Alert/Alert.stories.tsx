import type { Meta, StoryObj } from '@storybook/react'
import { Alert } from '@/src/components/Alert'

const meta: Meta<typeof Alert> = {
  title: 'Alert',
  component: Alert,
  argTypes: {
    type: { control: 'select', options: ['error', 'warning', 'info'] },
    message: { type: 'string' },
    iconName: { type: 'string' },
    displayIcon: { type: 'boolean' },
  },
}

export default meta

type Story = StoryObj<typeof Alert>

export const Warning: Story = {
  args: {
    type: 'warning',
    message: 'Proceed with caution',
    displayIcon: true,
  },
}

export const Error: Story = {
  args: {
    type: 'error',
    message: 'The transaction will most likely fail',
    displayIcon: true,
  },
}

export const Info: Story = {
  args: {
    type: 'info',
    message: 'This is info block',
    displayIcon: true,
  },
}
