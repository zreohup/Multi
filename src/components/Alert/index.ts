import { Alert, AlertType, AlertTitleStyled } from './Alert'

export { Alert, AlertType, AlertTitleStyled }
