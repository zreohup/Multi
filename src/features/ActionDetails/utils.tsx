import { TransactionDetails } from '@safe-global/store/gateway/AUTO_GENERATED/transactions'
import { ActionValueDecoded, AddressInfoIndex } from '@safe-global/store/gateway/types'
import { getActionName } from '../TransactionActions/components/TxActionsList'
import { Badge } from '@/src/components/Badge'
import { CircleProps, View, Text } from 'tamagui'
import { ListTableItem } from '../ConfirmTx/components/ListTable'
import { Logo } from '@/src/components/Logo'
import { ellipsis } from '@/src/utils/formatters'
import { CopyButton } from '@/src/components/CopyButton'
import { Identicon } from '@/src/components/Identicon'
import { Address } from '@/src/types/address'
import { EthAddress } from '@/src/components/EthAddress'
import { EncodedData } from '@/src/components/EncodedData/EncodedData'
import { SafeFontIcon } from '@/src/components/SafeFontIcon'

const badgeProps: CircleProps = { borderRadius: '$2', paddingHorizontal: '$2', paddingVertical: '$1' }

type formatActionDetailsReturn = {
  txData: TransactionDetails['txData']
  action: ActionValueDecoded
}

const getContractCall = (action: ActionValueDecoded, addressInfoIndex?: AddressInfoIndex) => {
  return addressInfoIndex?.[action.to]
}

const TxOptions = ({ value }: { value: string }) => {
  return (
    <View flexDirection="row" alignItems="center">
      <CopyButton value={value} color={'$textSecondaryLight'} />
      <SafeFontIcon name="external-link" size={14} color="$textSecondaryLight" />
    </View>
  )
}

const getContractItemLayout = ({
  logoUri,
  value,
  name,
}: {
  logoUri?: string | null
  value: string
  name?: string | null
}) => ({
  label: 'Contract',
  render: () => (
    <View flexDirection="row" alignItems="center" gap="$2">
      {logoUri ? <Logo logoUri={logoUri} size="$6" /> : <Identicon address={value as Address} size={24} />}
      <Text fontSize="$4">{ellipsis(name || value, 16)}</Text>
      <TxOptions value={value} />
    </View>
  ),
})

export const formatActionDetails = ({ txData, action }: formatActionDetailsReturn): ListTableItem[] => {
  if (!txData) {
    return []
  }

  let columns: ListTableItem[] = []

  if (action.dataDecoded?.method) {
    columns.push({
      label: 'Call',
      render: () => (
        <Badge
          circleProps={badgeProps}
          themeName="badge_background"
          fontSize={12}
          circular={false}
          content={getActionName(action)}
        />
      ),
    })
  } else {
    columns.push({
      label: 'Interacted with',
      render: () => (
        <View flexDirection="row" alignItems="center" gap="$2">
          <Identicon address={action.to as Address} size={24} />
          <EthAddress copy copyProps={{ color: '$textSecondaryLight' }} address={action.to as Address} />
          <SafeFontIcon name="external-link" size={14} color="$textSecondaryLight" />
        </View>
      ),
    })
  }

  const contractCall = getContractCall(action, txData.addressInfoIndex as AddressInfoIndex)

  if (contractCall) {
    columns.push(getContractItemLayout(contractCall))
  } else if (action.to) {
    columns.push(getContractItemLayout({ value: action.to }))
  }

  if (action.dataDecoded) {
    columns = [
      ...columns,
      ...action.dataDecoded.parameters.map((param) => ({
        label: param.name,
        render: () => {
          if (param.type === 'address') {
            return <EthAddress copy copyProps={{ color: '$textSecondaryLight' }} address={param.value as Address} />
          }

          return <Text>{param.value}</Text>
        },
      })),
    ]
  } else if (action.data) {
    columns = [
      ...columns,
      {
        label: 'Data',
        render: () => <EncodedData data={action.data} />,
      },
    ]
  }

  return columns
}
