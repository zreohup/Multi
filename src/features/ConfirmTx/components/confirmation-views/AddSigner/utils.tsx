import { SafeFontIcon } from '@/src/components/SafeFontIcon'
import { Logo } from '@/src/components/Logo'
import { ellipsis } from '@/src/utils/formatters'
import { Text, View } from 'tamagui'

import { MultisigExecutionDetails } from '@safe-global/store/gateway/AUTO_GENERATED/transactions'
import { shortenAddress } from '@safe-global/utils/utils/formatters'
import { Chain } from '@safe-global/store/gateway/AUTO_GENERATED/chains'
import { Identicon } from '@/src/components/Identicon'
import { NormalizedSettingsChangeTransaction } from '../../ConfirmationView/types'
import { CopyButton } from '@/src/components/CopyButton'
import { TouchableOpacity } from 'react-native'

export const getSignerName = (txInfo: NormalizedSettingsChangeTransaction) => {
  if (!txInfo.settingsInfo) {
    return ''
  }

  const newSigner = 'owner' in txInfo.settingsInfo && txInfo.settingsInfo.owner

  if (!newSigner) {
    return ''
  }

  return newSigner.name ? ellipsis(newSigner.name, 18) : shortenAddress(newSigner.value)
}

export const formatAddSignerItems = (
  txInfo: NormalizedSettingsChangeTransaction,
  chain: Chain,
  executionInfo: MultisigExecutionDetails,
  viewOnExplorer: () => void,
) => {
  const newSignerAddress = getSignerName(txInfo)

  const items = [
    {
      label: 'New signer',
      render: () => (
        <View flexDirection="row" alignItems="center" gap="$2">
          <Identicon address={txInfo.settingsInfo?.owner?.value} size={24} />
          <Text fontSize="$4">{newSignerAddress}</Text>
          <CopyButton value={txInfo.settingsInfo?.owner?.value} color={'$textSecondaryLight'} />

          <TouchableOpacity onPress={viewOnExplorer}>
            <SafeFontIcon name="external-link" size={14} color="$textSecondaryLight" />
          </TouchableOpacity>
        </View>
      ),
    },
    {
      label: 'Network',
      render: () => (
        <View flexDirection="row" alignItems="center" gap="$2">
          <Logo logoUri={chain.chainLogoUri} size="$6" />
          <Text fontSize="$4">{chain.chainName}</Text>
        </View>
      ),
    },
  ]

  const hasThresholdChanged = txInfo.settingsInfo?.threshold !== executionInfo.confirmationsRequired
  if (hasThresholdChanged) {
    items.push({
      label: 'Threshold change',
      render: () => (
        <View flexDirection="row" alignItems="center" gap="$2">
          <Text fontSize="$4">
            {txInfo.settingsInfo?.threshold}/{executionInfo.signers.length}
          </Text>
          <Text textDecorationLine="line-through" color="$textSecondaryLight" fontSize="$4">
            {executionInfo.confirmationsRequired}/{executionInfo.signers.length}
          </Text>
        </View>
      ),
    })
  }

  return items
}
