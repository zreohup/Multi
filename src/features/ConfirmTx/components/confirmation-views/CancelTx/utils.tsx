import { Logo } from '@/src/components/Logo'
import { Text, View } from 'tamagui'
import { Chain } from '@safe-global/store/gateway/AUTO_GENERATED/chains'

export const formatCancelTxItems = (chain: Chain) => {
  return [
    {
      label: 'Network',
      render: () => (
        <View flexDirection="row" alignItems="center" gap="$2">
          <Logo logoUri={chain.chainLogoUri} size="$6" />
          <Text fontSize="$4">{chain.chainName}</Text>
        </View>
      ),
    },
  ]
}
