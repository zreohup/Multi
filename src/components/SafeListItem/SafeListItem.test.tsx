import { render } from '@/src/tests/test-utils'
import { SafeListItem } from '.'
import { Text, View } from 'tamagui'
import { ellipsis } from '@/src/utils/formatters'
import { Alert } from '../Alert'

describe('SafeListItem', () => {
  it('should render the default markup', () => {
    const { getByText } = render(
      <SafeListItem label="A label" leftNode={<Text>Left node</Text>} rightNode={<Text>Right node</Text>} />,
    )

    expect(getByText('A label')).toBeTruthy()
    expect(getByText('Left node')).toBeTruthy()
    expect(getByText('Right node')).toBeTruthy()
  })

  it('should render a list item, with type and icon', () => {
    const { getByText, getByTestId } = render(
      <SafeListItem
        label="A label"
        type="some type"
        icon="add-owner"
        leftNode={<Text>Left node</Text>}
        rightNode={<Text>Right node</Text>}
      />,
    )

    expect(getByText('A label')).toBeTruthy()
    expect(getByText('some type')).toBeTruthy()
    expect(getByTestId('safe-list-add-owner-icon')).toBeTruthy()
    expect(getByText('Left node')).toBeTruthy()
    expect(getByText('Right node')).toBeTruthy()
  })

  it('should render a list item with truncated label when the label text length is very long', () => {
    const text = 'A very long label text to test if it it will truncate, in this case it should truncate.'
    const { getByText, getByTestId } = render(
      <SafeListItem label={text} type="some type" icon="add-owner" leftNode={<Text>Left node</Text>} />,
    )

    expect(getByText(ellipsis(text, 30))).toBeTruthy()
    expect(getByText('some type')).toBeTruthy()
    expect(getByTestId('safe-list-add-owner-icon')).toBeTruthy()
    expect(getByText('Left node')).toBeTruthy()
  })

  it('should render a list item with a custom label template', () => {
    const container = render(
      <SafeListItem
        label={
          <View>
            <Text>Here is my label</Text>
          </View>
        }
        type="some type"
        icon="add-owner"
        leftNode={<Text>Left node</Text>}
      />,
    )

    expect(container.getByText('Here is my label')).toBeTruthy()
    expect(container.getByText('some type')).toBeTruthy()
    expect(container.getByTestId('safe-list-add-owner-icon')).toBeTruthy()
    expect(container.getByText('Left node')).toBeTruthy()

    expect(container).toMatchSnapshot()
  })

  it('should render bottomContent when provided', () => {
    const { getByText } = render(
      <SafeListItem
        label="A label"
        leftNode={<Text>Left node</Text>}
        bottomContent={
          <View testID="bottom-content">
            <Text>Bottom content text</Text>
          </View>
        }
      />,
    )

    expect(getByText('A label')).toBeTruthy()
    expect(getByText('Left node')).toBeTruthy()
    expect(getByText('Bottom content text')).toBeTruthy()
  })

  it('should not render bottomContent when not provided', () => {
    const { queryByTestId } = render(<SafeListItem label="A label" leftNode={<Text>Left node</Text>} />)

    // Since bottomContent is not provided, there should be no bottom content container
    expect(queryByTestId('bottom-content')).toBeNull()
  })

  it('should render bottomContent with proper styling and layout', () => {
    const container = render(
      <SafeListItem
        label="A label"
        leftNode={<Text>Left node</Text>}
        bottomContent={
          <View testID="bottom-content-container">
            <Text testID="bottom-warning">Transaction warning message</Text>
          </View>
        }
      />,
    )

    expect(container.getByText('A label')).toBeTruthy()
    expect(container.getByTestId('bottom-content-container')).toBeTruthy()
    expect(container.getByText('Transaction warning message')).toBeTruthy()

    expect(container).toMatchSnapshot()
  })

  it('should render bottomContent with Alert component', () => {
    const { getByText } = render(
      <SafeListItem
        label="Transaction checks"
        leftNode={<Text>Shield icon</Text>}
        bottomContent={<Alert type="warning" message="Contract Changes Detected!" info="Review Details First" />}
      />,
    )

    expect(getByText('Transaction checks')).toBeTruthy()
    expect(getByText('Shield icon')).toBeTruthy()
    expect(getByText('Contract Changes Detected!')).toBeTruthy()
    expect(getByText('Review Details First')).toBeTruthy()
  })
})

describe('SafeListItem.Header', () => {
  it('should render the default markup', () => {
    const { getByText } = render(<SafeListItem.Header title="any title for your header here" />)

    expect(getByText('any title for your header here')).toBeTruthy()
  })
})
