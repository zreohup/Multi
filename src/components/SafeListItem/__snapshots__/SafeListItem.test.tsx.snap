// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SafeListItem should render a list item with a custom label template 1`] = `
<View>
  <View
    style={
      {
        "alignItems": "flex-start",
        "backgroundColor": "#FFFFFF",
        "borderBottomLeftRadius": 7,
        "borderBottomRightRadius": 7,
        "borderTopLeftRadius": 7,
        "borderTopRightRadius": 7,
        "flexDirection": "column",
        "flexWrap": "wrap",
        "gap": 12,
        "justifyContent": "flex-start",
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 12,
            "maxWidth": "100%",
          }
        }
      >
        <Text
          style={
            {
              "color": "#121312",
              "fontFamily": "DM Sans",
            }
          }
          suppressHighlighting={true}
        >
          Left node
        </Text>
        <View>
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "gap": 4,
                "marginBottom": 4,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#A1A3A7",
                    "fontSize": 10,
                  },
                  undefined,
                  {
                    "fontFamily": "SafeIcons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
              testID="safe-list-add-owner-icon"
            >
              î¤€
            </Text>
            <Text
              style={
                {
                  "color": "#A1A3A7",
                  "fontFamily": "DM Sans",
                  "fontSize": 13,
                  "marginBottom": 2,
                }
              }
              suppressHighlighting={true}
            >
              some type
            </Text>
          </View>
          <View>
            <Text
              style={
                {
                  "color": "#121312",
                  "fontFamily": "DM Sans",
                }
              }
              suppressHighlighting={true}
            >
              Here is my label
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`SafeListItem should render bottomContent with proper styling and layout 1`] = `
<View>
  <View
    style={
      {
        "alignItems": "flex-start",
        "backgroundColor": "#FFFFFF",
        "borderBottomLeftRadius": 7,
        "borderBottomRightRadius": 7,
        "borderTopLeftRadius": 7,
        "borderTopRightRadius": 7,
        "flexDirection": "column",
        "flexWrap": "wrap",
        "gap": 12,
        "justifyContent": "flex-start",
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 12,
            "maxWidth": "100%",
          }
        }
      >
        <Text
          style={
            {
              "color": "#121312",
              "fontFamily": "DM Sans",
            }
          }
          suppressHighlighting={true}
        >
          Left node
        </Text>
        <View>
          <Text
            style={
              {
                "color": "#121312",
                "fontFamily": "DMSans-SemiBold",
                "fontSize": 14,
              }
            }
            suppressHighlighting={true}
          >
            A label
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "marginTop": 12,
          "width": "100%",
        }
      }
    >
      <View
        testID="bottom-content-container"
      >
        <Text
          style={
            {
              "color": "#121312",
              "fontFamily": "DM Sans",
            }
          }
          suppressHighlighting={true}
          testID="bottom-warning"
        >
          Transaction warning message
        </Text>
      </View>
    </View>
  </View>
</View>
`;
